tests/unit/mesh_dp_api_tests.cc

```diff
@@ -0,0 +1,1490 @@
+#include <gtest/gtest.h>
+#include "mesh_dp_internal.h"
+#include <bsd/string.h>
+
+/**
+ * Test creation and deletion of a mesh client
+ */
+TEST(APITests_MeshClient, Test_CreateDeleteClient) {
+    MeshClient *mc = NULL;
+    int err;
+
+    err = mesh_create_client(&mc, NULL);
+    ASSERT_EQ(err, 0) << mesh_err2str(err);
+    ASSERT_NE(mc, (MeshClient *)NULL);
+
+    err = mesh_delete_client(&mc);
+    EXPECT_EQ(err, 0) << mesh_err2str(err);
+    EXPECT_EQ(mc, (MeshClient *)NULL);
+}
+
+/**
+ * Test negative scenario of creating a mesh client
+ */
+TEST(APITests_MeshClient, TestNegative_CreateClient) {
+    int err = mesh_create_client(NULL, NULL);
+    ASSERT_EQ(err, -MESH_ERR_BAD_CLIENT_PTR) << mesh_err2str(err);
+}
+
+/**
+ * Test negative scenario of deteling a mesh client
+ */
+TEST(APITests_MeshClient, TestNegative_DeleteClient) {
+    int err = mesh_delete_client(NULL);
+    ASSERT_EQ(err, -MESH_ERR_BAD_CLIENT_PTR) << mesh_err2str(err);
+}
+
+/**
+ * Create a mock mesh connection
+ */
+mcm_conn_context * mock_create_connection(mcm_conn_param* param)
+{
+    mcm_conn_context *conn;
+    
+    conn = (mcm_conn_context *)calloc(1, sizeof(mcm_conn_context));
+    conn->frame_size = 1024; /* Magic number for checking frame size */
+
+    return conn;
+}
+
+/**
+ * Destroy a mock mesh connection
+ */
+void mock_destroy_connection(mcm_conn_context *pctx)
+{
+    free(pctx);
+}
+
+/**
+ * Variable to hold the last timeout value passed to any API function
+ */
+int __last_timeout;
+
+/**
+ * Dequeue a mock mesh buffer
+ */
+mcm_buffer * mock_dequeue_buf(mcm_conn_context *pctx, int timeout, int *error_code)
+{
+    mcm_buffer *buf;
+
+    *error_code = 0;
+    __last_timeout = timeout;
+
+    /* Magic number for simulating that the connection is closed */
+    if (timeout == 12345)
+        return NULL;
+
+    buf = (mcm_buffer *)calloc(1, sizeof(mcm_buffer));
+    if (buf) {
+        buf->len = 192;         /* Magic number for checking buf data len */
+        pctx->frame_size = 384; /* Magic number for checking buf size */
+    }
+
+    return buf;
+}
+
+/**
+ * Enqueue a mock mesh buffer
+ */
+int mock_enqueue_buf(mcm_conn_context *pctx, mcm_buffer *buf)
+{
+    free(buf);
+    return 0;
+}
+
+/**
+ * APITests setup
+ */
+void APITests_Setup()
```

Comment:
Please take a look at https://github.com/meekrosoft/fff and consider using it. This would make it much easier to check things like whether the mock function has been called.

Reply:
Will integrate that in the next unit test extension cycle.